File: scripts/populate_leaderboard.js
// Script para poblar el leaderboard con datos de usuarios existentes en Firebase
const admin = require('firebase-admin');
const serviceAccount = require('../assets/data/serviceAccountKey.json');

// Inicializar Firebase Admin
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});

const db = admin.firestore();

async function populateLeaderboard() {
  console.log('🚀 Iniciando población del leaderboard...');
  
  try {
    // Obtener todos los usuarios
    console.log('📊 Obteniendo usuarios existentes...');
    const usersSnapshot = await db.collection('users').get();
    
    if (usersSnapshot.empty) {
      console.log('⚠️  No se encontraron usuarios en la base de datos');
      return;
    }
    
    console.log(`👥 Encontrados ${usersSnapshot.size} usuarios`);
    
    let processedUsers = 0;
    let successfulUpdates = 0;
    
    // Procesar cada usuario
    for (const userDoc of usersSnapshot.docs) {
      try {
        const userData = userDoc.data();
        const userId = userDoc.id;
        const username = userData.username || `Usuario ${userId.substring(0, 6)}`;
        
        // Calcular puntuación usando la misma lógica que LeaderboardService
        const level = userData.level || 1;
        const experience = userData.experience || 0;
        const battlesWon = userData.battlesWon || 0;
        const correctAnswers = userData.correctAnswers || 0;
        const completedMissions = userData.completedMissions ? userData.completedMissions.length : 0;
        
        const score = (level * 1000) + 
                     (experience * 10) + 
                     (battlesWon * 500) + 
                     (correctAnswers * 100) + 
                     (completedMissions * 200);
        
        // Solo actualizar si el usuario tiene algún progreso
        if (score > 1000) { // Más que solo el nivel inicial
          // Crear entrada en el leaderboard
          await db.collection('leaderboard').doc(userId).set({
            userId: userId,
            username: username,
            score: score,
            lastUpdated: admin.firestore.FieldValue.serverTimestamp()
          }, { merge: true });
          
          console.log(`✅ Usuario actualizado: ${username} (Puntuación: ${score})`);
          successfulUpdates++;
        } else {
          console.log(`⏭️  Usuario omitido: ${username} (Sin progreso significativo)`);
        }
        
        processedUsers++;
        
        // Pequeña pausa para evitar sobrecargar Firestore
        await new Promise(resolve => setTimeout(resolve, 100));
        
      } catch (e) {
        console.log(`❌ Error procesando usuario ${userDoc.id}: ${e.message}`);
      }
    }
    
    console.log('\n🎉 Proceso completado:');
    console.log(`   📊 Usuarios procesados: ${processedUsers}`);
    console.log(`   ✅ Actualizaciones exitosas: ${successfulUpdates}`);
    console.log(`   ❌ Errores: ${processedUsers - successfulUpdates}`);
    
    // Mostrar top 10 del leaderboard
    console.log('\n🏆 Top 10 del leaderboard:');
    const leaderboardSnapshot = await db
      .collection('leaderboard')
      .orderBy('score', 'desc')
      .limit(10)
      .get();
    
    leaderboardSnapshot.docs.forEach((doc, index) => {
      const data = doc.data();
      const username = data.username || 'Usuario desconocido';
      const score = data.score || 0;
      const medal = index === 0 ? '🥇' : index === 1 ? '🥈' : index === 2 ? '🥉' : '  ';
      console.log(`   ${medal} ${index + 1}. ${username} - ${score} puntos`);
    });
    
  } catch (error) {
    console.error('💥 Error crítico:', error);
    process.exit(1);
  }
  
  console.log('\n✨ Script completado exitosamente');
  process.exit(0);
}

// Ejecutar el script
populateLeaderboard();

File: bin/populate_leaderboard.dart
import 'dart:io';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';
import '../lib/firebase_options.dart';

/// Script para poblar el leaderboard con datos de usuarios existentes
void main() async {
  print('🚀 Iniciando población del leaderboard...');
  
  try {
    // Inicializar Firebase
    await Firebase.initializeApp(
      options: DefaultFirebaseOptions.currentPlatform,
    );
    print('✅ Firebase inicializado correctamente');
    
    final FirebaseFirestore firestore = FirebaseFirestore.instance;
    
    // Obtener todos los usuarios
    print('📊 Obteniendo usuarios existentes...');
    QuerySnapshot usersSnapshot = await firestore.collection('users').get();
    
    if (usersSnapshot.docs.isEmpty) {
      print('⚠️  No se encontraron usuarios en la base de datos');
      return;
    }
    
    print('👥 Encontrados ${usersSnapshot.docs.length} usuarios');
    
    int processedUsers = 0;
    int successfulUpdates = 0;
    
    // Procesar cada usuario
    for (QueryDocumentSnapshot userDoc in usersSnapshot.docs) {
      try {
        final userData = userDoc.data() as Map<String, dynamic>;
        final userId = userDoc.id;
        final username = userData['username'] ?? 'Usuario ${userId.substring(0, 6)}';
        
        // Calcular puntuación usando la misma lógica que LeaderboardService
        final level = userData['level'] ?? 1;
        final experience = userData['experience'] ?? 0;
        final battlesWon = userData['battlesWon'] ?? 0;
        final correctAnswers = userData['correctAnswers'] ?? 0;
        final completedMissions = userData['completedMissions']?.length ?? 0;
        
        final score = (level * 1000) + 
                     (experience * 10) + 
                     (battlesWon * 500) + 
                     (correctAnswers * 100) + 
                     (completedMissions * 200);
        
        // Solo actualizar si el usuario tiene algún progreso
        if (score > 1000) { // Más que solo el nivel inicial
          // Crear entrada en el leaderboard
          await firestore.collection('leaderboard').doc(userId).set({
            'userId': userId,
            'username': username,
            'score': score,
            'lastUpdated': FieldValue.serverTimestamp(),
          }, SetOptions(merge: true));
          
          print('✅ Usuario actualizado: $username (Puntuación: $score)');
          successfulUpdates++;
        } else {
          print('⏭️  Usuario omitido: $username (Sin progreso significativo)');
        }
        
        processedUsers++;
        
        // Pequeña pausa para evitar sobrecargar Firestore
        await Future.delayed(Duration(milliseconds: 100));
        
      } catch (e) {
        print('❌ Error procesando usuario ${userDoc.id}: $e');
      }
    }
    
    print('\n🎉 Proceso completado:');
    print('   📊 Usuarios procesados: $processedUsers');
    print('   ✅ Actualizaciones exitosas: $successfulUpdates');
    print('   ❌ Errores: ${processedUsers - successfulUpdates}');
    
    // Mostrar top 10 del leaderboard
    print('\n🏆 Top 10 del leaderboard:');
    QuerySnapshot leaderboardSnapshot = await firestore
        .collection('leaderboard')
        .orderBy('score', descending: true)
        .limit(10)
        .get();
    
    for (int i = 0; i < leaderboardSnapshot.docs.length; i++) {
      final doc = leaderboardSnapshot.docs[i];
      final data = doc.data() as Map<String, dynamic>;
      final username = data['username'] ?? 'Usuario desconocido';
      final score = data['score'] ?? 0;
      final medal = i == 0 ? '🥇' : i == 1 ? '🥈' : i == 2 ? '🥉' : '  ';
      print('   $medal ${i + 1}. $username - $score puntos');
    }
    
  } catch (e, stackTrace) {
    print('💥 Error crítico: $e');
    print('Stack trace: $stackTrace');
    exit(1);
  }
  
  print('\n✨ Script completado exitosamente');
  exit(0);
}